### PR TITLE
🌱 Add compare util using go-cmp, modify webhooks & KCP controller

### LINKED ISSUE #8186: Consider replacing reflect.DeepEqual
I think we should consider replacing reflect.DeepEqual. Whilte it works well to diff two structs, it's very very had to debug what exactly is not equal. This makes it very hard to produce good logs to tell users why something was considered not equal.

I think go-cmp might be a good alternative. For more details, see: https://github.com/google/go-cmp

We already use go-cmp today to produce a diff in test code after reflect.DeepEqual only told us that something is not equal. The comparision algorithm of go-cmp seems to be better and additionally we get the beneift of getting a result which tells us where the difference is. This would be super valuable for our prod code to tell folks why we are doing things. 

Using cmp.Diff for logs while we use reflect.DeepEqual for the actual comparison is not great as there are small differences, e.g. unexported fields.

One example where I think this is super valuable is in the KCP controller where we could tell folks why exactly we are triggering a MachineRollout and not just: well reflect.DeepEqual told us that there is a difference.

There are a few more places like e.g. the ClusterClass controller where we compare variable definitions where we could have a log message (with high v) that could log exactly why the definitions are different.

(xref: https://github.com/kubernetes-sigs/cluster-api/pull/8107#discussion_r1109601666)

/kind feature

### PR DESCRIPTION
**Pull request by sbueringer**
**What this PR does / why we need it**:
Needs a bit more work and some manual tests (want to see how the diffs are looking in logs)

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #8186

Example logs for KCP controller
```
I0516 16:01:22.008912      58 controller.go:412] "Rolling out Control Plane machines: Machine capi-quickstart-control-plane-xvq7h needs rollout: Machine KubeadmConfig ClusterConfiguration is outdated: diff: &v1beta1.ClusterConfiguration{\n    ... // 3 identical fields\n    KubernetesVersion:    \"\",\n    ControlPlaneEndpoint: \"\",\n    APIServer: v1beta1.APIServer{\n      ControlPlaneComponent: {},\n      CertSANs: []string{\n        \"localhost\",\n-       \"127.0.0.1\",\n        \"0.0.0.0\",\n        \"host.docker.internal\",\n      },\n      TimeoutForControlPlane: nil,\n    },\n    ControllerManager: {ExtraArgs: {\"enable-hostpath-provisioner\": \"true\"}},\n    Scheduler:         {},\n    ... // 5 identical fields\n  }" controller="kubeadmcontrolplane" controllerGroup="controlplane.cluster.x-k8s.io" controllerKind="KubeadmControlPlane" KubeadmControlPlane="default/capi-quickstart-control-plane" namespace="default" name="capi-quickstart-control-plane" reconcileID="93e4bce7-00da-44c7-9421-a714ecf996ca" Cluster="default/capi-quickstart" machinesNeedingRollout=["capi-quickstart-control-plane-xvq7h"]
```

Message after removing escaping
```
Rolling out Control Plane machines: Machine capi-quickstart-control-plane-xvq7h needs rollout: Machine KubeadmConfig ClusterConfiguration is outdated: diff: &v1beta1.ClusterConfiguration{
    ... // 3 identical fields
    KubernetesVersion:    "",
    ControlPlaneEndpoint: "",
    APIServer: v1beta1.APIServer{
      ControlPlaneComponent: {},
      CertSANs: []string{
        "localhost",
-       "127.0.0.1",
        "0.0.0.0",
        "host.docker.internal",
      },
      TimeoutForControlPlane: nil,
    },
    ControllerManager: {ExtraArgs: {"enable-hostpath-provisioner": "true"}},
    Scheduler:         {},
    ... // 5 identical fields
  }
```

Example for webhook
```
k apply -f /var/folders/v0/mny6w2dd33gfzwmwxb3wyyx40000gq/T/kubectl-edit-3932073547.yaml
[kubeadmcontrolplanetemplate.controlplane.cluster.x-k8s.io/quick-start-control-plane-2](http://kubeadmcontrolplanetemplate.controlplane.cluster.x-k8s.io/quick-start-control-plane-2) configured
[kubeadmcontrolplanetemplate.controlplane.cluster.x-k8s.io/quick-start-control-plane-3](http://kubeadmcontrolplanetemplate.controlplane.cluster.x-k8s.io/quick-start-control-plane-3) configured
The KubeadmControlPlaneTemplate "quick-start-control-plane" is invalid: spec.template.spec: Invalid value: v1beta1.KubeadmControlPlaneTemplate{TypeMeta:v1.TypeMeta{Kind:"KubeadmControlPlaneTemplate", APIVersion:"[controlplane.cluster.x-k8s.io/v1beta1](http://controlplane.cluster.x-k8s.io/v1beta1)"}, ObjectMeta:v1.ObjectMeta{Name:"quick-start-control-plane", GenerateName:"", Namespace:"default", SelfLink:"", UID:"288374d2-7a3b-4a26-9606-0627bd06db1d", ResourceVersion:"2293", Generation:2, CreationTimestamp:time.Date(2024, time.May, 16, 16, 6, 12, 0, time.Local), DeletionTimestamp:<nil>, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string{"[kubectl.kubernetes.io/last-applied-configuration](http://kubectl.kubernetes.io/last-applied-configuration)":"{\"apiVersion\":\"[controlplane.cluster.x-k8s.io/v1beta1](http://controlplane.cluster.x-k8s.io/v1beta1)\",\"kind\":\"KubeadmControlPlaneTemplate\",\"metadata\":{\"annotations\":{},\"creationTimestamp\":\"2024-05-16T16:06:12Z\",\"generation\":1,\"name\":\"quick-start-control-plane\",\"namespace\":\"default\",\"resourceVersion\":\"2293\",\"uid\":\"288374d2-7a3b-4a26-9606-0627bd06db1d\"},\"spec\":{\"template\":{\"metadata\":{},\"spec\":{\"kubeadmConfigSpec\":{\"clusterConfiguration\":{\"apiServer\":{\"certSANs\":[\"localhost\",\"0.0.0.0\",\"host.docker.internal\"]},\"controllerManager\":{\"extraArgs\":{\"enable-hostpath-provisioner\":\"true\"}},\"dns\":{},\"etcd\":{},\"networking\":{},\"scheduler\":{}},\"files\":[{\"content\":\"SOMETESTCONTENT\\n\",\"path\":\"abc\"}],\"format\":\"cloud-config\",\"initConfiguration\":{\"localAPIEndpoint\":{},\"nodeRegistration\":{\"criSocket\":\"/var/run/containerd/containerd.sock\",\"imagePullPolicy\":\"IfNotPresent\",\"kubeletExtraArgs\":{\"eviction-hard\":\"nodefs.available\\u003c0%,nodefs.inodesFree\\u003c0%,imagefs.available\\u003c0%\"}}},\"joinConfiguration\":{\"discovery\":{},\"nodeRegistration\":{\"criSocket\":\"/var/run/containerd/containerd.sock\",\"imagePullPolicy\":\"IfNotPresent\",\"kubeletExtraArgs\":{\"eviction-hard\":\"nodefs.available\\u003c0%,nodefs.inodesFree\\u003c0%,imagefs.available\\u003c0%\"}}}},\"rolloutStrategy\":{\"rollingUpdate\":{\"maxSurge\":1},\"type\":\"RollingUpdate\"}}}}}\n"}, OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ManagedFields:[]v1.ManagedFieldsEntry{v1.ManagedFieldsEntry{Manager:"kubectl-client-side-apply", Operation:"Update", APIVersion:"[controlplane.cluster.x-k8s.io/v1beta1](http://controlplane.cluster.x-k8s.io/v1beta1)", Time:time.Date(2024, time.May, 16, 16, 7, 25, 0, time.Local), FieldsType:"FieldsV1", FieldsV1:(*v1.FieldsV1)(0x4000a540a8), Subresource:""}}}, Spec:v1beta1.KubeadmControlPlaneTemplateSpec{Template:v1beta1.KubeadmControlPlaneTemplateResource{ObjectMeta:v1beta1.ObjectMeta{Labels:map[string]string(nil), Annotations:map[string]string(nil)}, Spec:v1beta1.KubeadmControlPlaneTemplateResourceSpec{MachineTemplate:(*v1beta1.KubeadmControlPlaneTemplateMachineTemplate)(nil), KubeadmConfigSpec:v1beta1.KubeadmConfigSpec{ClusterConfiguration:(*v1beta1.ClusterConfiguration)(0x4000d862c0), InitConfiguration:(*v1beta1.InitConfiguration)(0x40000ea620), JoinConfiguration:(*v1beta1.JoinConfiguration)(0x4001172000), Files:[]v1beta1.File{v1beta1.File{Path:"abc", Owner:"", Permissions:"", Encoding:"", Append:false, Content:"SOMETESTCONTENT\n", ContentFrom:(*v1beta1.FileSource)(nil)}}, DiskSetup:(*v1beta1.DiskSetup)(nil), Mounts:[]v1beta1.MountPoints(nil), PreKubeadmCommands:[]string(nil), PostKubeadmCommands:[]string(nil), Users:[]v1beta1.User(nil), NTP:(*v1beta1.NTP)(nil), Format:"cloud-config", Verbosity:(*int32)(nil), UseExperimentalRetryJoin:false, Ignition:(*v1beta1.IgnitionSpec)(nil)}, RolloutBefore:(*v1beta1.RolloutBefore)(nil), RolloutAfter:<nil>, RolloutStrategy:(*v1beta1.RolloutStrategy)(0x4000a54150), RemediationStrategy:(*v1beta1.RemediationStrategy)(nil)}}}}: KubeadmControlPlaneTemplate spec.template.spec field is immutable. Please create new resource instead. Diff: v1beta1.KubeadmControlPlaneTemplateResourceSpec{
    MachineTemplate: nil,
    KubeadmConfigSpec: v1beta1.KubeadmConfigSpec{
      ClusterConfiguration: &v1beta1.ClusterConfiguration{
        ... // 3 identical fields
        KubernetesVersion:    "",
        ControlPlaneEndpoint: "",
        APIServer: v1beta1.APIServer{
          ControlPlaneComponent: {},
          CertSANs: []string{
            "localhost",
-           "127.0.0.1",
            "0.0.0.0",
            "host.docker.internal",
          },
          TimeoutForControlPlane: nil,
        },
        ControllerManager: {ExtraArgs: {"enable-hostpath-provisioner": "true"}},
        Scheduler:         {},
        ... // 5 identical fields
      },
      InitConfiguration: &{NodeRegistration: {CRISocket: "/var/run/containerd/containerd.sock", KubeletExtraArgs: {"eviction-hard": "nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%"}, ImagePullPolicy: "IfNotPresent"}},
      JoinConfiguration: &{NodeRegistration: {CRISocket: "/var/run/containerd/containerd.sock", KubeletExtraArgs: {"eviction-hard": "nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%"}, ImagePullPolicy: "IfNotPresent"}},
      ... // 11 identical fields
    },
    RolloutBefore: nil,
    RolloutAfter:  nil,
    ... // 2 identical fields
  }
```

### CODE REVIEW

**Comment by sbueringer** [comment]
Not sure how to flag this correctly

File: controlplane/kubeadm/internal/filters_test.go [hunk 1]
```diff
@@ -450,7 +478,29 @@ func TestMatchInitOrJoinConfiguration(t *testing.T) {
...
+        Taints:    nil,
+        ... // 3 identical fields
+      },
+      LocalAPIEndpoint: {},
+      SkipPhases:       nil,
+      Patches:          nil,
+    },
+    JoinConfiguration: nil,
+    Files:             nil,
+    ... // 10 identical fields
```
Comment (chrischdi) [inline controlplane/kubeadm/internal/filters_test.go:502]: Note: Not sure how good these diffing is, because we want to not have a match.

The diff will change as soon as we add fields (which should in practice normally never happen :-) )

TLDR: ok to do it like this, let the future us handle this if it happens :D

File: internal/util/compare/equal.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+		if r := recover(); r != nil {
+			equal = false
+			if err, ok := r.(error); ok {
+				matchErr = err
+				return
+			}
+
+			if errMsg, ok := r.(string); ok {
+				matchErr = errors.New(errMsg)
+			}
```
Comment (chrischdi) [inline internal/util/compare/equal.go:40]: Should we add something like "unable to recover" in case of line 38 is `!ok`?


```suggestion
        return
			}
      
      matchErr = errors.New("unable to recover from cmp.Diff panic")
```

File: internal/util/compare/equal.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+		}
+	}()
+
+	diff = cmp.Diff(x, y)
+
+	if diff != "" {
+		// Replace non-breaking space (NBSP) through a regular space.
+		// This prevents output like this: "\u00a0\u00a0int(\n-\u00a0\t1,\n+\u00a0\t2,\n\u00a0\u00a0)\n"
+		diff = strings.ReplaceAll(diff, "\u00a0", " ")
+		// Replace \t through "  " because it's easier to rid in log output
```
Comment (chrischdi) [inline internal/util/compare/equal.go:50]:
```suggestion
		// Replace \t through "  " because it's easier to read in log output
```

**Review by chrischdi — COMMENTED** [review]
small nit

File: internal/util/compare/equal.go (hunk 2 above)
Comment (sbueringer) [reply internal/util/compare/equal.go:40]: Good catch. I wouldn't call this "not recovered", PTAL what I changed

File: controlplane/kubeadm/internal/filters_test.go (hunk 1 above)
Comment (sbueringer) [reply controlplane/kubeadm/internal/filters_test.go:502]: In general I like the format that cmp.Diff is producing per default. Theoretically we could implement a custom reporter, but I don't have a good idea how to do it better than they do :)

**Comment by chrischdi** [comment]
/lgtm

(failing test though)

**Comment by sbueringer** [comment]
Ups, thx, fixed

**Comment by sbueringer** [comment]
Had to rebase because of the defaulting webhook fix

/assign @fabriziopandini @chrischdi

**Review by fabriziopandini — COMMENTED** [review]
Nice
This will greatly improve debugging rollouts in the logs
/lgtm
/approved
